### PR TITLE
`FastAPI`: Add DTOs and pipeline run endpoint

### PR DESCRIPTION
**Pull request by MichaelOwenDyer**

### CODE REVIEW

File: app/web/routers/pipelines.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+router = APIRouter(prefix="/api/v1/pipelines", tags=["pipelines"])
+logger = logging.getLogger(__name__)
+
+
+@router.post("/tutor-chat/{variant}/run", status_code=status.HTTP_202_ACCEPTED)
+def run_pipeline(variant: str, dto: TutorChatPipelineExecutionDTO):
+    callback = TutorChatStatusCallback(run_id=dto.settings.authentication_token)
+    pipeline = TutorChatPipeline(callback=callback)
+    pipeline(dto=dto)
```
Comment (Hialus) [inline app/web/routers/pipelines.py:18]: I see 2 problems here:
1. You do not do any authentication to ensure that the incoming request is from Artemis. You can reuse the token auth from PyrisV1 here
2. You do not seem to start the pipeline in a separate thread, which would make this is blocking call, which we want to avod.

File: app/web/status/status_update.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        self.url = url
+
+    @abstractmethod
+    def on_status_update(self, status: Any):
+        pass
+
+
+class TutorChatStatusCallback(StatusCallback):
+    def __init__(self, run_id: str):
+        url = f"/api/v1/public/pyris/tutorchat/runs/{run_id}/status"
```
Comment (Hialus) [inline app/web/status/status_update.py:22]: This URL is missing the protocol & domain in front. This should be extracted from the settings DTO that Artemis gives you.

File: .gitignore [hunk 3]
```diff
@@ -158,3 +158,4 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 .idea/
+/app/pipeline_playground.py
```
Comment (Hialus) [inline .gitignore:161]: Not sure if this should be added here. I'd rather opt for an entire folder or just add this to your local only git ignore

**Review by Hialus — CHANGES_REQUESTED** [review]

File: app/pipeline/chat/tutor_chat_pipeline.py [hunk 4]
```diff
@@ -18,43 +32,127 @@ class TutorChatPipeline(Pipeline):
...
+                StageDTO(
+                    name="Final Stage",
+                    weight=70,
+                    state=StageStateDTO.ERROR,
+                    message="Error running tutor chat pipeline",
+                )
+            )
+            status_dto = TutorChatStatusUpdateDTO(stages=stages)
+
+        self.callback.on_status_update(status_dto)
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/pipeline/chat/tutor_chat_pipeline.py:158]: We will do our own end to end testing after we deploy the application.

File: README.MD [hunk 5]
```diff
@@ -1 +1,16 @@
...
+
+### Run server
+ - Run server: 
+   ```[bash]
+     APPLICATION_YML_PATH=<path-to-your-application-yml-file> LLM_CONFIG_PATH=<path-to-your-llm-config-yml> uvicorn app.main:app --reload
+    ```
+ - Access API docs: http://localhost:8000/docs
+
+## With docker
+TBD
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply README.MD:16]: Docker setup is yet to be merged. It waits this pull request to be merged. After it's merged, we will have a more clear picture regarding docker setup instructions.

File: app/pipeline/chat/tutor_chat_pipeline.py [hunk 6]
```diff
@@ -18,43 +39,240 @@ class TutorChatPipeline(Pipeline):
...
+        self.llm = IrisLangchainChatModel(
+            request_handler=request_handler, completion_args=completion_args
         )
-        # Create the pipeline
-        self.pipeline = prompt | llm | StrOutputParser()
+        self.callback = callback
+
+        # Create the pipelines
+        self.file_selector_pipeline = FileSelectorPipeline()
+        self.pipeline = self.llm | StrOutputParser()
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/pipeline/chat/tutor_chat_pipeline.py:58]: For this pipeline we currently use gpt35. We do not use environment variables for default model names.

File: app/pipeline/chat/tutor_chat_pipeline.py [hunk 7]
```diff
@@ -18,43 +39,240 @@
...
+        stages.append(
+            StageDTO(
+                name="Response generation",
+                weight=10,
+                state=StageStateDTO.DONE,
+                message="Generated response",
+            )
+        )
+        self._send_status_update(stages, response)
+
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/pipeline/chat/tutor_chat_pipeline.py]: Addressed in the next commit.

File: app/common/custom_exceptions.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            },
+        )
+
+
+class PipelineNotFoundException(HTTPException):
+    def __init__(self):
+        super().__init__(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail={
+                "type": "not_found",
```
Comment (Hialus) [inline app/common/custom_exceptions.py:42]: Please be more specific with the types. E.g. "pipeine_not_found", so Artemis can clearly distingush between them

**Comment by Hialus** [comment]
str | None in the DTOs to Optional. See also https://stackoverflow.com/questions/69440494/python-3-10-optionaltype-or-type-none

**Comment by Hialus** [comment]
Merge MessageDTO and IrisMessage into one PyrisMessage class

File: app/domain/health_status_dto.py
Comment (Hialus) [inline app/domain/health_status_dto.py:1]: Remove all health related stuff

File: app/web/data/repository_fetch.py
Comment (kaancayli) [inline app/web/data/repository_fetch.py:1]: Delete this.

File: app/web/routers/health.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+from fastapi import APIRouter, status, Response
+
+router = APIRouter(prefix="/api/v1/health", tags=["health"])
+
+
+@router.get("/")
+def health_check():
```
Comment (kaancayli) [inline app/web/routers/health.py:7]: Returon 200 and do auth check

File: app/web/routers/models.py
Comment (kaancayli) [inline app/web/routers/models.py:1]: Delete this.

**Comment by Hialus** [comment]
Comments will be adressed in future PRs.

**Pull request merged**